src/Shared/HttpValidationProblemDetailsJsonConverter.cs

```diff
@@ -60,10 +61,26 @@ public static void WriteProblemDetails(Utf8JsonWriter writer, HttpValidationProb
         writer.WriteStartObject();
         ProblemDetailsJsonConverter.WriteProblemDetails(writer, value, options);
 
+        static string ConvertName(string name, JsonSerializerOptions options)
+        {
+            if (string.IsNullOrEmpty(name) || options?.DictionaryKeyPolicy == null)
+            {
+                return name;
+            }
+
+            var tokens = name.Split(propertySeparator);
+            for (var i = 0; i < tokens.Length; i++)
+            {
+                tokens[i] = options!.DictionaryKeyPolicy.ConvertName(tokens[i]);
+            }
+
+            return string.Join(propertySeparator, tokens);
+        }
+
         writer.WriteStartObject(Errors);
```

Comment:
An alternative is to replace this section with a call `JsonSerializer.Serialize(writer, Errors, options)`. That way we're no longer in the business of serializing dictionaries.

Reply:
I really do prefer that. Let me give a try because I was trying to keep the code with the minimal changes

Reply:
But i think we will end up with the same issue because of the `.`

Reply:
It might be worthwhile trying to make that a S.T.J feature (a new dictionary policy)? The error keys are entirely opaque to the framework, so trying to interpret them as Json paths or .NET object paths by default feels a bit problematic.

Reply:
@pranavkm , i didn't get exactly what you mean about S.T.J feature, sorry about that. The idea is not interpreting the keys as Json Paths, what I am doing is just splitting the text, convert each word to camelCase and join it again. Basically, string manipulation

Reply:
I updated the code to use the JsonSerializer directly, instead of the for + string manipulation, but as expected the results are not right:

Expected:
```json
{
  "title": "One or more validation errors occurred.",
  "status": 400,
  "errors": {
    "property": [
      "error0"
    ],
    "twoWords": [
      "error1"
    ],
    **"topLevelProperty.propertyName"**: [
      "error2"
    ]
  }
}

```
Actual result:
```json
{
  "title": "One or more validation errors occurred.",
  "status": 400,
  "errors": {
    "property": [
      "error0"
    ],
    "twoWords": [
      "error1"
    ],
    **"topLevelProperty.PropertyName"**: [
      "error2"
    ]
  }
}
```

Reply:
I'm not sure it's a common enough use case to warrant a S.T.J feature. Do you know of anywhere else we would use it?

Reply:
Sorry, I'm having a hard time figuring out why we're messing with these keys. Let's say I do this:

```C#
new ValidationProblemDetails
{
   Errors: 
   { 
      ["3.2.Alpha"] = "Some Error"
    }
}
```

Why are we choosing to rewrite the key? Keys are entirely opaque to the framework - they may happen to look like .NET member accessors, but that doesn't mean we should arbitrarily transform them. If System.Text.Json had a feature that said transform all keys to have an arbitrary format and the user configured it feels correct because you opted in. I'm having a hard time trying to see why we would try to do something special here, and why doing that is pertinent.

Reply:
> Why are we choosing to rewrite the key? Keys are entirely opaque to the framework - they may happen to look like .NET member accessors, but that doesn't mean we should arbitrarily transform them.

I think this is a reasonable argument for not adding this feature to System.Text.Json. However, based on the amount of feedback on the issue, this seems to be clearly the default behavior customers want from `ValidationProblemDetails.Errors` when using camelCased JSON properties.

The original issue mentioned nothing about the `DictionaryKeyPolicy`. To most people, the fact that the `ValidationProblemDetails.Errors` is a dictionary is merely an implementation detail. They just want the error keys to look like JSON member accessors. Later comments merely point out that a `DictionaryKeyPolicy` couldn't even be used to work around the current behavior.

I don't we should be using `DictionaryKeyPolicy` to gate this behavior especially if we're going to camelCase after every `.` meaning we're not really following any built-in `DictionaryKeyPolicy` to begin with. It would probably be better to continue to ignore the `DictionaryKeyPolicy` in here and gate this on the `PropertyNamingPolicy` instead. It might be *technically* wrong, but it behaves the way people want it to. That said, it would make this change more breaking.

https://github.com/dotnet/aspnetcore/issues/17999#issuecomment-810913391

Reply:
Is using the DictionaryKeyPolicy but not changing it to do more an option? Users could definitely modify their app to achieve the "right" behavior. I imagine the easiest way for them would be to override `ControllerBase.ValidationProblem` to tweak how the errors dictionary is populated.

Reply:
You both know the source code way more than me and I am probably missing some important facts, so, with this disclaimer :) I would say that based on the issue the full camelCase (after every `.`) is the only behavior accepted and I totally agree that use the `DictionaryKeyPolicy` may not be the best option since we are looking to a behavior not covered by the CamelCase policy.

Also, using the `ControllerBase.ValidationProblem` could be a workaround for users, on the other hand, in my opinion they already had one (rewrite the Converter https://github.com/dotnet/aspnetcore/issues/17999#issuecomment-598241620) that can be applied global instead of changing every controller class to inherit from a new class or override the method every time.

I know that may have already too many options, but one suggestion is not use any of the `System.Text.Json.JsonSerializerOptions`(`DictionaryKeyPolicy` or `PropertyNamingPolicy`) and instead just add a new option to the `Microsoft.AspNetCore.Mvc.JsonOptions`, something like:

``` c#
services.AddControllers().AddJsonOptions(options => 
 {
    options.ModelStateErrorsCamelCase = true
 });
```

Reply:
Just to add an information about the `S.T.J.JsonCamelCaseNamingPolicy` the fact that the second part, after the `.` in my example, is not getting converted to camel case does not have anything related to the `.`. Basically, the method has the following condition that will `break` the `for` operation and skip the next words conversion.

https://github.com/dotnet/runtime/blob/69d75858e0d0d9577b286e0b2aeabc62cd3e383e/src/libraries/System.Text.Json/Common/JsonCamelCaseNamingPolicy.cs#L36
``` c#
if (i == 1 && !char.IsUpper(chars[i]))
{
   break;
}
```

Reply:
With that, I think updating the code to using the built-in S.T.J feature and not do the full camel case might be the most correct approach to be consistent with the rest of the framework even if that is not exactly the behavior expected by customers but in this case it will, at least, not be expected anywhere we are converting to JSON a dictionary in ASP.NET.

@pranavkm/ @halter73, is that make sense?